File: C++/Hard/Accepted/best-time-to-buy-and-sell-stock-iii_8.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        
        int n = prices.size();
        vector<int> dp(n, 0);
        int minPrice = prices[0];
        int maxProfit = 0;
        for (int i = 1; i < n; ++i) {
            minPrice = min(minPrice, prices[i]);
            maxProfit = max(maxProfit, prices[i] - minPrice);
            dp[i] = maxProfit;
        }
        int maxPrice = prices[n - 1];
        maxProfit = 0;
        int ans = 0;
        for (int i = n - 2; i >= 0; --i) {
            maxPrice = max(maxPrice, prices[i]);
            maxProfit = max(maxProfit, maxPrice - prices[i]);
            ans = max(ans, dp[i] + maxProfit);
        }
        return ans;
    }
};

File: Rust/Hard/Accepted/best-time-to-buy-and-sell-stock-iii_8.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        
        let mut dp = vec![vec![0; 5]; prices.len()];
        for i in 0..prices.len() {
            for j in 0..5 {
                dp[i][j] = match j {
                    0 => 0,
                    1 => {
                        if i == 0 {
                            -prices[i]
                        } else {
                            std::cmp::max(dp[i - 1][1], -prices[i])
                        }
                    }
                    2 => {
                        if i == 0 {
                            0
                        } else {
                            std::cmp::max(dp[i - 1][2], dp[i - 1][1] + prices[i])
                        }
                    }
                    3 => {
                        if i == 0 {
                            -prices[i]
                        } else {
                            std::cmp::max(dp[i - 1][3], dp[i - 1][2] - prices[i])
                        }
                    }
                    4 => {
                        if i == 0 {
                            0
                        } else {
                            std::cmp::max(dp[i - 1][4], dp[i - 1][3] + prices[i])
                        }
                    }
                    _ => unreachable!(),
                }
            }
        }
        dp[prices.len() - 1][4]
    }
}